zcash_client_sqlite/src/wallet/common.rs

```diff
@@ -225,3 +225,55 @@ where
         .filter_map(|r| r.transpose())
         .collect::<Result<_, _>>()
 }
+
+pub(crate) fn count_outputs(
+    conn: &rusqlite::Connection,
+    account: AccountId,
+    min_value: NonNegativeAmount,
+    exclude: &[ReceivedNoteId],
+    protocol: ShieldedProtocol,
+) -> Result<usize, rusqlite::Error> {
+    let (table_prefix, _, _) = per_protocol_names(protocol);
+
+    let excluded: Vec<Value> = exclude
+        .iter()
+        .filter_map(|ReceivedNoteId(p, n)| {
+            if *p == protocol {
+                Some(Value::from(*n))
+            } else {
+                None
+            }
+        })
+        .collect();
+    let excluded_ptr = Rc::new(excluded);
+
+    conn.query_row(
+        &format!(
+            "SELECT COUNT(*)
+             FROM {table_prefix}_received_notes
+             INNER JOIN accounts
+               ON accounts.id = {table_prefix}_received_notes.account_id
+             WHERE value >= :min_value
```

Comment:
Unlike `get_notes` or `select_spendable_notes`, this counts notes in unmined transactions. Is that intended? It doesn't take into account whether the transactions are likely to be mined. In general it makes me slightly more nervous of metadata leakage when transaction construction policy is dependent on mempool information (because that's less consistent across nodes).

To exclude notes in unmined transactions you would need something like:
```suggestion
             FROM {table_prefix}_received_notes rn
             INNER JOIN accounts ON accounts.id = rn.account_id
             INNER JOIN transactions ON transactions.id_tx = rn.tx
             WHERE transactions.block IS NOT NULL
             AND value >= :min_value
```

Reply:
Oh wait, is `commitment_tree_position` always NULL if the note is in an unmined transaction? I don't see how we could know it otherwise. In that case why does `get_notes` explicitly join on `transactions` and check for `transactions.block IS NOT NULL`? (Since `get_notes` also has `commitment_tree_position IS NOT NULL`, checking `transactions.block IS NOT NULL` there should be redundant by the same argument.)

Reply:
Hm, I think the `commitment_tree_position` check is actually superfluous; it's the join against the `blocks` table that ensures that the transaction has been mined.